fix(ivy): a collection of ngcc fixes for processing various 3rd party packages

petebacondarwin opened the pull request:
The good news is that with this PR I think we can process all the packages reported in #29564 (except for primeng).

Further, in order to process `@angular/fire` we need to provide the following file in your project root:

**ngcc.config.js**

```js
module.exports = {
  packages: {
    '@angular/fire': {
      entryPoints: {
        './database-deprecated': {
          override: {
            'main': '../bundles/database-deprecated.umd.js',
          }
        },
      },
    },
  },
};
```

This is because the `database-deprecated` and `database` entry-points both point to the same source files. This is a bug in the package that should/will be fixed at the source.

Fixes #31069

petebacondarwin committed:
feat(ivy): customize ngcc via configuration files

There are scenarios where it is not possible for ngcc to guess the format
or configuration of an entry-point just from the files on disk.

Such scenarios include:

1) Unwanted entry-points: A spurious package.json makes ngcc think
there is an entry-point when there should not be one.

2) Deep-import entry-points: some packages allow deep-imports but do not
provide package.json files to indicate to ngcc that the imported path is
actually an entry-point to be processed.

3) Invalid/missing package.json properties: For example, an entry-point
that does not provide a valid property to a required format.

The configuration is provided by one or more `ngcc.config.js` files:

* If placed at the root of the project, this file can provide configuration
for named packages (and their entry-points) that have been npm installed
into the project.

* If published as part of a package, the file can provide configuration
for entry-points of the package.

The configured of a package at the project level will override any
configuration provided by the package itself.

petebacondarwin committed:
test(ivy): ngcc - remove use of mock-fs in tests

Now that ngcc uses a `FileSystem` throughout we no longer need
to rely upon mocking out the real file-system with mock-fs.

petebacondarwin committed:
test(ivy): ngcc refactor mock file-systems to make each spec independent

Previously each test relied on large shared mock file-systems, which
makes it difficult to reason about what is actually being tested.

This commit breaks up these big mock file-systems into smaller more
focused chunks.

gkalpak commented on packages/compiler-cli/ngcc/src/packages/configuration.ts:
```diff
@@ -0,0 +1,120 @@
+/**
+ * Configuration options for an entry-point.
+ *
+ * The existence of a configuration for a path tells ngcc that this is should be
```
_is should be_ --> _should be_

gkalpak commented on packages/compiler-cli/ngcc/src/packages/configuration.ts:
```diff
@@ -0,0 +1,120 @@
+  }
+
+  private loadPackageConfig(packagePath: AbsoluteFsPath): NgccPackageConfig {
+    const configFilePath = AbsoluteFsPath.join(packagePath, NGCC_CONFIG_FILENAME);
```
I find it a bit confusing that we use the same filename for both project and package configs. Why is that?

petebacondarwin replied:
I guess it never occurred to me to call them something different... They are used in different places so it is their position that distinguishes them rather than their filename. Interested in @alxhub's view on this. I could be convinced to change it to `ngcc.config.js` for the project and `ngcc.package.config.js`??

gkalpak commented on packages/compiler-cli/ngcc/src/packages/configuration.ts:
```diff
@@ -0,0 +1,120 @@
+  private evalSrcFile(srcPath: AbsoluteFsPath): any {
+    const src = this.fs.readFile(srcPath);
+    const sandbox = {module: {exports: {}}};
+    vm.runInNewContext(src, sandbox);
```
What is the intention of using this (instead of directly requireing the code)?

petebacondarwin replied:
We want to use the `FileSystem` for all disk interactions and `require()` would undermine this. So this was the best way I could find to load the source file via the `FileSystem` and evaluate it.

petebacondarwin committed:
fix(ivy): ngcc - infer entry-point typings from format paths

Some packages do not actually provide a `typings` field in their
package.json. But TypeScript naturally infers the typings file from
the location of the JavaScript source file.

This commit modifies ngcc to do a similar inference when finding
entry-points to process.

Fixes #28603 (FW-1299)

petebacondarwin committed:
fix(ivy): ngcc - support bare array constructor param decorators

Previously we expected the constructor parameter `decorators`
property to be an array wrapped in a function. Now we also support
an array not wrapped in a function.

petebacondarwin committed:
refactor(ivy): ngcc - pass whole entry-point object to `makeEntryPointBundle()`

This simplifies the interface somewhat but also allows us to make use of
other properties of the EntryPoint object in the future.

petebacondarwin committed:
refactor(ivy): ngcc - expose the `entryPoint` from the `EntryPointBundle` interface

This will allow users of the `EntryPointBundle` to use some of the `EntryPoint`
properties without us having to pass them around one by one.

petebacondarwin committed:
refactor(ivy): ngcc - no  need to pass `isCore` explicitly

It is part of `EntryPointBundle` so we can just use that, which
is generally already passed around.

petebacondarwin committed:
refactor(ivy): ngcc - pass `bundle` to `DecorationAnalyzer`

Rather than passing a number of individual arguments, we can
just pass an `EntryPointBundle`, which already contains them.

This is also a precursor to using more of the properties in the bundle.

petebacondarwin committed:
fix(ivy): ngcc - do not analyze files outside the current package

Our module resolution prefers `.js` files over `.d.ts` files because
occasionally libraries publish their typings in the same directory
structure as the compiled JS files, i.e. adjacent to each other.

The standard TS module resolution would pick up the typings
file and add that to the `ts.Program` and so they would be
ignored by our analyzers. But we need those JS files, if they
are part of the current package.

But this meant that we also bring in JS files from external
imports from outside the package, which is not desired.
This was happening for the `@fire/storage` enty-point
that was importing the `firebase/storage` path.

In this commit we solve this problem, for the case of imports
coming from a completely different package, by saying that any
file that is outside the package root directory must be an external
import and so we do not analyze those files.

This does not solve the potential problem of imports between
secondary entry-points within a package but so far that does
not appear to be a problem.

JoostK reviewed (approved):
This looks great! I love the cleanup commits! 👏

gkalpak commented on packages/compiler-cli/src/ngtsc/path/src/types.ts:
```diff
@@ -66,26 +69,34 @@ export const AbsoluteFsPath = {
+    if (sf.fileName !== AbsoluteFsPath.from(sf.fileName).toString()) {
+      throw new Error(`Bad file cast: ${sf.fileName} -> ${AbsoluteFsPath.from(sf.fileName)}`);
+    }
+    return sf.fileName as any as AbsoluteFsPath;
```
Since you are already calling `AbsoluteFsPath.from()`, you can avoid the casting:

```ts
const absPath = AbsoluteFsPath.from(sf.fileName);
if (sf.fileName !== absPath.toString()) {
  throw new Error(`Bad file cast: ${sf.fileName} -> ${absPath}`);
}
return absPath;
```

gkalpak commented on packages/compiler-cli/src/ngtsc/path/src/types.ts:
```diff
@@ -66,26 +69,34 @@ export const AbsoluteFsPath = {
-      AbsoluteFsPath { return AbsoluteFsPath.fromUnchecked(path.posix.join(basePath, ...paths));},
+  join: function(basePath: AbsoluteFsPath, ...paths: Array<PathSegment|AbsoluteFsPath|string>):
+      AbsoluteFsPath {
+        return path.posix.join(basePath as any as string, ...(paths as string[])) as any;
```
FWIW, allowing strings does not guarantee that the returned path will be properly formatted. E.g. `path.posix.join('/base/path', 'foo\\bar')` returns `/base/path/foo\\bar`.

petebacondarwin replied:
Gadzooks! What a mess! So we still need to normalize the slashes throughout. I always thought that `join()` would also normalize.

gkalpak commented on packages/compiler-cli/src/ngtsc/path/src/types.ts:
```diff
@@ -66,26 +69,34 @@ export const AbsoluteFsPath = {
+      },
 
   /**
    * Wrapper around `path.resolve` that returns an absolute paths.
```
_an absolute paths_ --> _an absolute path_ / _absolute paths_

gkalpak commented on packages/compiler-cli/src/ngtsc/path/src/types.ts:
```diff
@@ -103,20 +114,28 @@ export const PathSegment = {
+export const ModuleSpecifier = {
+  from: function(specifier: string): ModuleSpecifier {
+    // This could be a relative PathSegment or an AbsoluteFsPath.
+    return specifier as any;
```
Shouldn't we do some normalization anyway (e.g. replace `\\` with `/`)?

petebacondarwin replied:
Yes you are probably right :-)

gkalpak commented on packages/compiler-cli/src/ngtsc/path/src/types.ts:
```diff
@@ -10,27 +10,35 @@ import * as ts from 'typescript';
+ * to convert.
  */
-export type AbsoluteFsPath = BrandedPath<'AbsoluteFsPath'>;
+export interface AbsoluteFsPath extends String { _brand: 'AbsoluteFsPath'; }
```
Is it intentional that these types are not assignable to `string` (even if they extend `String`)?
It would make more sense to be to be able to use them as `string`, without having to cast them or call `.toString()`.

petebacondarwin replied:
Yes it is intentional. I think I explain this in the commit message. The idea is that when we are passing string back to 3rd party libraries (i.e. node.js or TS) then we have to consciously cast it to a string (or call `toString()`. This is to ensure that we really are aware of when these things are being used - and also gives us a way to search for places where we are exiting the branded paths ecosystem.

gkalpak replied:
I am not convinced the benefits are worth the extra work, but I am not convinced they aren't, so happy to go with that :grin:

petebacondarwin commented on packages/compiler-cli/src/ngtsc/path/src/logical.ts:
```diff
@@ -30,11 +28,12 @@ export const LogicalProjectPath = {
   relativePathBetween: function(from: LogicalProjectPath, to: LogicalProjectPath): PathSegment {
-    let relativePath = path.posix.relative(path.posix.dirname(from), to);
+    let relativePath =
+        path.posix.relative(path.posix.dirname(from as any as string), to as any as string);
```
I really don't think it matters in a situation where you are immediately casting to something else and it is fewer characters :-)

AFAICT the only benefit of casting to `unknown` is that you can't then use it for anything until you have actually cast it to something tangible, whereas `any` lets you do whatever you like.

Carniatto commented:
Do we have any updates on this PR ?

petebacondarwin commented:
Hi @Carniatto - sorry this is taking a while. It is currently blocked on https://github.com/angular/angular/pull/30921, which I hope to land soon.

Carniatto commented:
@petebacondarwin No problem, is there any workaround that I can use for now? I did wish to keep using ivy in the meanwhile

petebacondarwin commented:
@Carniatto - are you able to build a local version of Angular? If so then I could rebase this and you could patch Angular locally with the PR.

alxhub reviewed (approved):
I rebased this, updated paths, and changed the `fix(ivy): configuration...` commit to a `feat` commit.

alxhub commented:
[Presubmit](http://test/OCL:255076239:BASE:255077600:1561506645388:5294fee)

[Ivy Presubmit](http://test/OCL:255076239:BASE:255077958:1561506758646:8a5c1fcf)

kara commented:
@alxhub Thanks!
